code.js

```diff
@@ -1,3 +1,53 @@
+class PriorityQueue {
+    constructor() {
+        this.queue = [];
+    }
+
+    enqueue(vertex, priority) {
+        this.queue.push({ vertex, priority });
+        this.queue.sort((a, b) => a.priority - b.priority);
+    }
+
+    dequeue() {
+        return this.queue.shift();
+    }
+
+    isEmpty() {
+        return this.queue.length === 0;
+    }
+}
+
 function dijkstra(graph, sourceNode) {
-    return [];
+    const distances = {};
+    const previous = {};
+    const pq = new PriorityQueue();
+
+    for (let vertex in graph) {
+        distances[vertex] = Infinity;
+        previous[vertex] = null;
+    }
+    distances[sourceNode] = 0;
+
+    for (let vertex in graph) {
+        pq.enqueue(vertex, distances[vertex]);
+    }
+
+    while (!pq.isEmpty()) {
+        let minNode = pq.dequeue();
+        let currentVertex = minNode.vertex;
+
+        for (let neighbor in graph[currentVertex]) {
+            let weight = graph[currentVertex][neighbor];
+            let alt = distances[currentVertex] + weight;
+            if (alt < distances[neighbor]) {
+                distances[neighbor] = alt;
+                previous[neighbor] = currentVertex;
+                pq.enqueue(neighbor, alt);
```

Comment:
Would this not lead to nodes being in the queue multiple times?

Reply:
I accidentally pushed node_modules which i meant to keep locally so i fixed that and previously had sources in readme right above the runtime complexity. Could you please explain what you are getting at with this comment, "Would this not lead to nodes being in the queue multiple times?" thank you

Reply:
You already have all nodes in the queue, right? So here you're adding the same node again, although possibly with a different distance estimate, as far as I can see.